compiler/qsc_partial_eval/src/lib.rs

```diff
@@ -0,0 +1,711 @@
+        let eval_result = state.eval(
+            self.package_store,
+            &mut scope.env,
+            &mut self.backend,
+            &mut receiver,
+            &[],
+            StepAction::Continue,
+        );
+        match eval_result {
+            Ok(step_result) => {
+                let StepResult::Return(value) = step_result else {
+                    panic!("evaluating a classical expression should always return a value");
+                };
+                self.eval_context
+                    .get_current_scope_mut()
+                    .insert_expr_value(expr_id, value);
+            }
+            Err((error, _)) => self.error = Some(Error::EvaluationFailed(error)),
+        };
+    }
+
+    fn eval_classical_stmt(&mut self, stmt_id: StmtId) {
+        let current_package_id = self.get_current_package_id();
+        let store_stmt_id = StoreStmtId::from((current_package_id, stmt_id));
+        let stmt = self.package_store.get_stmt(store_stmt_id);
+        let scope_exec_graph = self.get_current_scope_exec_graph().clone();
+        let scope = self.eval_context.get_current_scope_mut();
+        let exec_graph = exec_graph_section(&scope_exec_graph, stmt.exec_graph_range.clone());
+        let mut out = Vec::new();
+        let mut receiver = GenericReceiver::new(&mut out);
+        let mut state = State::new(current_package_id, exec_graph, None);
+        _ = state.eval(
+            self.package_store,
+            &mut scope.env,
+            &mut self.backend,
+            &mut receiver,
+            &[],
+            StepAction::Continue,
+        );
+    }
+
+    fn generate_expr_call(&mut self, callee_expr_id: ExprId, args_expr_id: ExprId) {
+        let (store_item_id, functor_app, callable_decl) = self.eval_callee_expr(callee_expr_id);
+
+        // We generate instructions differently depending on whether
+        match &callable_decl.implementation {
+            CallableImpl::Intrinsic => {
+                self.generate_expr_call_intrinsic(store_item_id, &callable_decl, args_expr_id);
+            }
+            CallableImpl::Spec(spec_impl) => {
+                self.generate_expr_call_spec(store_item_id, functor_app, spec_impl, args_expr_id);
+            }
+        };
+    }
+
+    fn generate_expr_call_intrinsic(
+        &mut self,
+        store_item_id: StoreItemId,
+        callable_decl: &CallableDecl,
+        args_expr_id: ExprId,
+    ) {
+        // Check if the callable is already in the program, and if not add it.
+        let callable_name = callable_decl.name.name.to_string();
+        #[allow(clippy::map_entry)]
+        if !self.callables_map.contains_key(&callable_name) {
+            let callable = self.create_intrinsic_callable(store_item_id, callable_decl);
+            let callable_id = self.assigner.next_callable();
+            self.program.callables.insert(callable_id, callable);
+            self.callables_map.insert(callable_name, callable_id);
+        }
```

Comment:
you can actually avoid the clippy allow if you use `Entry::Vacant` and some intentional ordering of the lines inside an if-let (note you'll need to add a `use` of `std::collections::hash_map::Entry`):

```suggestion
        if let Entry::Vacant(entry) = self.callables_map.entry(callable_name) {
            let callable_id = self.assigner.next_callable();
            entry.insert(callable_id);
            let callable = self.create_intrinsic_callable(store_item_id, callable_decl);
            self.program.callables.insert(callable_id, callable);
        }
```

Reply:
I see. I opted to allow this clippy lint because I thought the "logical" order of operations is to insert the callable ID to the map until after the callable has been added to the program, but since that conflicted with the borrow checker, I decided to not use `entry`. However, it is not a big deal the order in which these operations happen as long as they happen "at the same time" before we need to check again.

Long way of saying, I'll update it to use your suggestion 😄

Reply:
Updated.